src/main/java/com/google/sps/data/DatastoreService.java

```diff
@@ -0,0 +1,208 @@
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.Query;
+import com.google.cloud.datastore.QueryResults;
+//import com.google.cloud.datastore.StructuredQuery.OrderBy;
+import java.util.ArrayList;
+import java.util.List;
+
+public class DatastoreService {
+
+    /* Returns a List of all Users stored in Datastore.
+        !!! Used for debugging.
+    */
+    public List<User> getAllUsers() {
+        Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
+        Query<Entity> query = Query.newEntityQueryBuilder().setKind("User").build();
+        QueryResults<Entity> results = datastore.run(query);
+        List<User> users = new ArrayList<>();
+        while (results.hasNext()) {
+            Entity entity = results.next();
+
+            String username = entity.getString("username");
+            String email = entity.getString("email");
+            String passwordHash = entity.getString("passwordHash");
+
+            String firstName = entity.getString("firstName");
+            String middleName = entity.getString("middleName");
+            String lastName = entity.getString("lastName");
+            
+            var calendar = entity.getList("calendar");
+            var journal = entity.getList("journal");
+            
+            long panicButtonPressed = entity.getLong("panicButtonPressed");
+
+            User user = new User(username, email, passwordHash, firstName, middleName, lastName, calendar, journal, panicButtonPressed);
+            users.add(user);
+        }
+        return users;
+    }
+    
+    /* Validates user credentials. */
+    public boolean validateCredentials(User user, String password) {
+        if (user == null) {
+            return false;
+        }
+        return user.getPasswordHash().equals(password);
+    }
+
+    /* Returns a User with the given username. */
+    public User getUser(String username) {
+        Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
+        Query<Entity> query = Query.newEntityQueryBuilder()
+            .setKind("User")
+            .setFilter(PropertyFilter.eq("username", username))
+            .build();
+        QueryResults<Entity> results = datastore.run(query);
+        if (results.hasNext()) {
+            Entity entity = results.next();
+
+            String email = entity.getString("email");
+            String passwordHash = entity.getString("passwordHash");
+
+            String firstName = entity.getString("firstName");
+            String middleName = entity.getString("middleName");
+            String lastName = entity.getString("lastName");
+
+            var calendar = entity.getList("calendar");
+            var journal = entity.getList("journal");
+
+            long panicButtonPressed = entity.getLong("panicButtonPressed");
+
+            User user = new User(username, email, passwordHash, firstName, middleName, lastName, calendar, journal, panicButtonPressed);

```

Comment:
Refactor this into a method (i.e. createUser)? You can replace line 28-41 with this method too to reduce duplicate code.

Reply:
Done! :D 
Wrote two methods to keep the code a bit cleaner, thanks!